### DISCUSSION TITLE
`vercel.json` rewrites not working as expected

### DISCUSSION THREAD

**Discussion by surjithctly** [vercel/community#698] [Help] [answered]
Can anyone tell me how to configure vercel.json properly? I have added a rewrites but its not working. 

I have added two domains in the domains section:

mainsite.com
newsite.dev

Now whenever, user visits newsite.dev, the homepage should render /components from the mainsite.com

Here's the vercel.json i added in the root of the project

```json
{
    "rewrites": [{
        "source": "/:path*",
        "has": [{
            "type": "host",
            "value": "newsite.dev"
        }],
        "destination": "/components/:path*"
    }]
}

```

Current result: I'm seeing the homepage while visiting both domains. am I missing something? How do i make sure vercel.json is taking effect?

Please help 🙏

**Comment by amyegan** [comment, marked as answer]
Hi @surjithctly. Please correct me if I'm misunderstanding you. It looks like you're trying to redirect traffic from `newsite.dev` to `mainsite.com` from the config of `mainsite.com`. 

The config for `mainsite` cannot control what happens to traffic on `newsite`. If you move the rewrite to your `newsite` config then it should work for you.

**Reply by surjithctly** [reply]
I have only one project, then I added both domains in vercel domains. So normally if I visit both domains, I get to see same page. Same as we have `.vercel.app` and custom domain. 

So there is only one source of truth. As per the doc, I understood vercel can track the hostname and rewrite based on it. (Same as how multi-tenant apps work). So if its coming from a host, serve some subfolder to that client. 

Hope it make sense now.

**Reply by surjithctly** [reply]
I also tried with middleware similar to https://github.com/vercel/platforms/blob/main/middleware.ts but it didn't work.

**Reply by amyegan** [reply]
Do you have `newsite.dev` setup in the Dashboard to redirect to the base URL? That could be overriding any middleware or other rewrites you try to set. If you do have this type of redirect set, try changing it to `No Redirect`. You can find this setting by going to Dashboard -> Project -> Settings -> Domains

<img width="639" alt="image" src="https://user-images.githubusercontent.com/8247105/177825222-9ed707cf-9f4f-49f5-b319-4fc5906aad7e.png">

**Reply by surjithctly** [reply]
Yes, I have given no redirect. But since it was not working, I made a redirect instead of a rewrite (from DNS). That solves my issue for now. 

`newsite.dev` redirects to `mainsite.com/folder`

**Comment by pfcodes** [comment]
This still doesn't work.

**Reply by amyegan** [reply]
Can you describe exactly what isn't working and what you have tried already?

**Reply by pfcodes** [reply]
Like @surjithctly, I just used a redirect instead of a rewrite. 

Simply put, this doesn't work on Vercel:
```
{
  source: '/:path*',
  has: [
    {
      type: 'host',
      value: 'hostname.com',
    },
  ],
  destination: '/page',
},
```

**Reply by amyegan** [reply]
I know others have had problems when the origin project already has its own index page. [Edge Middleware](https://vercel.com/docs/concepts/functions/edge-middleware#) is a better solution in this case. You can set a custom [matcher](https://vercel.com/docs/concepts/functions/edge-middleware#match-paths-based-on-custom-matcher-config) config or [conditional statements](https://vercel.com/docs/concepts/functions/edge-middleware#match-paths-based-on-conditional-statements) to target only specific paths. And [`@vercel/edge`](https://vercel.com/docs/concepts/functions/edge-functions/vercel-edge-package#) provides helper functions for things like rewrites and geolocation.

The [Hostname Rewrites using Edge Middleware](https://vercel.com/templates/next.js/hostname-rewrites) template is also a useful reference

**Comment by wallawe** [comment]
I'm running into this issue as well. @amyegan this feels like something that should be addressed or at least noted in the documentation if the bug isn't going to be addressed. This took me several hours of trying to fix before I found this issue.

**Reply by heikopaiko** [reply]
Same issue over here @amyegan. Even your support agent recommended me the snippet mentioned above, which is not working at all. 

What i am trying to achieve is to mask a subdomain to a certain path of the base.

{
  source: '/:path*',
  has: [
    {
      type: 'host',
      value: 'app.hostname.com',
    },
  ],
  destination: '/page/app',
},

any ideas on how to achieve that? I am lost atm :(

**Reply by amyegan** [reply]
There are a few limitations when using the basic redirects. If you are trying to redirect the root, and not just sub-paths, you'll need to add config for `source: '/'` as well. And the source shouldn't have it's own index.html file, since that would take priority over the redirect. This documentation is being updated, but I'll make a note to be sure we make these limitations clear.

I prefer [Edge Middleware rewrites](https://vercel.com/docs/concepts/edge-network/redirects#edge-middleware). If you're interested in that approach, we have a [Maintenance Page template](https://vercel.com/templates/next.js/maintenance-page) demonstrating how Middleware rewrites can be implemented with [Edge Config](https://vercel.com/docs/concepts/edge-network/edge-config) for more flexibility.
